### PR TITLE
#76726 OIOSAML 3.0: Update the .NET web sample on GitHub to support OIOSAML 3.0

### PR DESCRIPTION
**Pull request by trungdang194**
#76726 OIOSAML 3.0: Update the .NET web sample on GitHub to support OIOSAML 3.0
- Update Web.Config due to SigningCertificate schema changed and applied changed after comparing source https://github.com/digst/OIOSAML.Net and target branch https://tfs2019.globeteam.com/Safewhere%20Identify/3rd%20party%20source/_git/3rd%20party%20source
- Add external assemblies to support encrypt/decrypt with GCM

### CODE REVIEW

**Comment by thuannguy** [comment]
Some issues:
1. Please upgrade the project to .NET 4.8
2. Safewhere.IdentityProvider.Core and Safewhere.ServiceProvider dlls cannot be pushed to GitHub. They are licensed dlls. (which is why we need to use dk.nita.saml20
3.

File: Kombit.Samples.CH.WebsiteDemo/Web.config [hunk 1]
```diff
@@ -65,105 +67,138 @@
...
+      <att name="https://data.gov.dk/model/core/eid/age" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/cprUuid" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/dateOfBirth" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/person/pid" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/professional/uuid/persistent" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/professional/rid" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/professional/productionUnit" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/professional/seNumber" isRequired="false"/>
     </RequestedAttributes>
-    <NameIdFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:X509SubjectName</NameIdFormat>
```
Comment (thuannguy) [inline Kombit.Samples.CH.WebsiteDemo/Web.config:74]: Kombit profile uses X509SubjectName. Please revert this change

File: Kombit.Samples.CH.WebsiteDemo/X509CertificateSigningSecurityTokenResolverFactory.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+using Safewhere.ServiceProvider.Util;
+using System;
+using System.IdentityModel.Selectors;
+
+namespace WebsiteDemo
+{
+    /// <summary>
+    /// A factory which returns X509CertificateSigningSecurityTokenResolverFactory for signing key resolver.
+    /// </summary>
+    public class X509CertificateSigningSecurityTokenResolverFactory : DefaultSecurityTokenResolverFactory
```
Comment (thuannguy) [inline Kombit.Samples.CH.WebsiteDemo/X509CertificateSigningSecurityTokenResolverFactory.cs]: The websitedemo project uses nita.saml2.0 to read SAML assertions. Does it need these resolvers? I thought it doesn't.

**Review by thuannguy — CHANGES_REQUESTED** [review]

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config [hunk 3]
```diff
@@ -1,61 +1,63 @@
...
   </location>
   <Federation auditLoggingType="dk.nita.saml20.Logging.Log4NetAuditLogger, dk.nita.saml20.ext.audit.log4net" xmlns="urn:dk.nita.saml20.configuration">
     <SessionTimeout>30</SessionTimeout>
-    <SigningCertificate findValue="4678237245FEDC8059D113675955DFB870D36BF4" storeLocation="LocalMachine" storeName="My" x509FindType="FindByThumbprint" />
+        <SigningCertificates>
+      <SigningCertificate findValue="21ad7d2d4280765bfe113b7dd5d62736c34e37bd" storeLocation="LocalMachine" storeName="My" x509FindType="FindByThumbprint"/>
+    </SigningCertificates>
     <AllowedAudienceUris>
-      <Audience>https://saml.claimapp.adgangsstyring.eksterntest-stoettesystemerne.dk</Audience>
+      <Audience>https://spdemo.gcm.safewhere.com</Audience>
```
Comment (thuannguy) [inline Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config]: Should keep the old url (even though we will update it with a new url later)

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config [hunk 4]
```diff
@@ -1,61 +1,63 @@
...
-      <add name="Redirect" type="dk.nita.saml20.Actions.RedirectAction, dk.nita.saml20" />
+      <clear/>
+      <add name="SetSamlPrincipal" type="dk.nita.saml20.Actions.SamlPrincipalAction, dk.nita.saml20"/>
+      <add name="LogAction" type="Kombit.Samples.CH.WebsiteDemo.LogAction, Kombit.Samples.CH.WebsiteDemo"/>
+      <add name="Redirect" type="dk.nita.saml20.Actions.RedirectAction, dk.nita.saml20"/>
     </Actions>
   </Federation>
   <SAML20Federation xmlns="urn:dk.nita.saml20.configuration">
     <ShowError>true</ShowError>
-    <ServiceProvider id="https://saml.claimapp.adgangsstyring.eksterntest-stoettesystemerne.dk" server="https://claimapp.adgangsstyring.eksterntest-stoettesystemerne.dk">
```
Comment (thuannguy) [inline Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config:54]: Should keep the old url (even though we will update it with a new url later)

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config [hunk 5]
```diff
@@ -65,105 +67,138 @@
       </md:ContactPerson>
     </ServiceProvider>
     <RequestedAttributes>
-      <att name="dk:gov:saml:attribute:AssuranceLevel" isRequired="true" />
-      <att name="dk:gov:saml:attribute:SpecVer" isRequired="true" />
-      <att name="dk:gov:saml:attribute:KombitSpecVer" isRequired="true" />
-      <att name="dk:gov:saml:attribute:CvrNumberIdentifier" isRequired="false" />
-      <att name="dk:gov:saml:attribute:Privileges_intermediate" isRequired="false" />
+      <att name="dk:gov:saml:attribute:AssuranceLevel" isRequired="true"/>
```
Comment (thuannguy) [inline Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config]: Please ask NGT for the required list. He has it ready now.

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config [hunk 6]
```diff
@@ -65,105 +67,138 @@
...
+      <att name="https://data.gov.dk/model/core/eid/dateOfBirth" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/person/pid" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/professional/uuid/persistent" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/professional/rid" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/professional/productionUnit" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/professional/seNumber" isRequired="false"/>
     </RequestedAttributes>
     <NameIdFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:X509SubjectName</NameIdFormat>
     <IDPEndPoints metadata="metadata">
-      <add id="https://saml.adgangsstyring.eksterntest-stoettesystemerne.dk">
```
Comment (thuannguy) [inline Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config]: Should keep the old url (even though we will update it with a new url later)

File: Deployment/Kombit.Samples.CH.WebsiteDemo/packages.config [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Kombit.Samples.BasicPrivilegeProfileParser" version="1.0.0.0" targetFramework="net45" />
+  <package id="Kombit.Samples.Common" version="1.0.0.2" targetFramework="net45" />
+  <package id="Microsoft.AspNet.WebApi.Client" version="5.2.3" targetFramework="net45" />
```
Comment (thuannguy) [inline Deployment/Kombit.Samples.CH.WebsiteDemo/packages.config]: Please upgrade to .NET 4.8 as well as latest version (5.2.7 or so)

Comment (thuannguy) [reply Deployment/Kombit.Samples.CH.WebsiteDemo/packages.config]: Same for other packages

File: Kombit.Samples.CH.WebsiteDemo/MyPage.aspx.cs [hunk 8]
```diff
@@ -75,7 +76,11 @@ protected void Btn_ReloginNoForceAuthnAssuranceLevel4_Click(object sender, Event
 
         private void HandleReloginNoForceAuthnAssuranceLevel(int assuranceLevel)
         {
-            SessionFactory.SessionContext.Current[SessionConstants.RequestedAssuranceLevel] = "urn:dk:gov:saml:attribute:AssuranceLevel:" + assuranceLevel;
+            var session = SessionStore.CurrentSession;
+            if (session != null)
+            {
+                SessionStore.CurrentSession[SessionConstants.RequestedAssuranceLevel] = "urn:dk:gov:saml:attribute:AssuranceLevel:" + assuranceLevel;
```
Comment (thuannguy) [inline Kombit.Samples.CH.WebsiteDemo/MyPage.aspx.cs]: This is not the right OIOSAML3 claim type. You can find the right one in the list that NGT will give you

File: Kombit.Samples.CH.WebsiteDemo/MyPage.aspx.cs [hunk 9]
```diff
@@ -117,7 +122,14 @@ protected static bool ValidateKombitAttributeProfile(Saml20Identity current)
             if (!current.HasAttribute("dk:gov:saml:attribute:CvrNumberIdentifier"))
```
Comment (thuannguy) [inline Kombit.Samples.CH.WebsiteDemo/MyPage.aspx.cs]: Need to use the new claim type

File: Kombit.Samples.CH.WebsiteDemo/Web.config [hunk 10]
```diff
@@ -65,105 +67,138 @@
...
+      <att name="https://data.gov.dk/model/core/eid/professional/orgName" isRequired="true"/>
+      
+      <att name="https://data.gov.dk/model/core/eid/bootstrapToken" isRequired="false"/>
+      <att name="https://data.gov.dk/concept/core/nsis/ial" isRequired="false"/>
+      <att name="https://data.gov.dk/concept/core/nsis/aal" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/fulName" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/firstName" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/lastName" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/email" isRequired="false"/>
+      <att name="https://data.gov.dk/model/core/eid/cprNumber" isRequired="false"/>
```
Comment (thuannguy) [inline Kombit.Samples.CH.WebsiteDemo/Web.config]: Same comments as those in the /deployment folder

**Review by thuannguy — CHANGES_REQUESTED** [review]

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config (hunk 3 above)
Comment (trungdang194) [reply Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config]: revert to the old url

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config (hunk 5 above)
Comment (trungdang194) [reply Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config]: update with new required list

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config (hunk 4 above)
Comment (trungdang194) [reply Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config:54]: reverted to the old url

File: Deployment/Kombit.Samples.CH.WebsiteDemo/packages.config (hunk 7 above)
Comment (trungdang194) [reply Deployment/Kombit.Samples.CH.WebsiteDemo/packages.config]: upgraded to .net 4.8 and updated packages to newer version

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config (hunk 6 above)
Comment (trungdang194) [reply Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config]: reverted to the old url

File: Kombit.Samples.CH.WebsiteDemo/MyPage.aspx.cs (hunk 8 above)
Comment (trungdang194) [reply Kombit.Samples.CH.WebsiteDemo/MyPage.aspx.cs]: changed to use NSIS assurance level claim https://data.gov.dk/concept/core/nsis/loa

File: Kombit.Samples.CH.WebsiteDemo/MyPage.aspx.cs (hunk 9 above)
Comment (trungdang194) [reply Kombit.Samples.CH.WebsiteDemo/MyPage.aspx.cs]: updated with OIOSAML 3 required claim types

File: Kombit.Samples.CH.WebsiteDemo/Web.config (hunk 10 above)
Comment (trungdang194) [reply Kombit.Samples.CH.WebsiteDemo/Web.config]: fixed

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config [hunk 11]
```diff
@@ -81,12 +86,21 @@
     </IDPEndPoints>
     <CommonDomain enabled="false" localReaderEndpoint="https://pfs04/demo/cdcreader.ashx" />
   </SAML20Federation>
+  <!--
+    For a description of web.config changes see http://go.microsoft.com/fwlink/?LinkId=235367.
+
+    The following attributes can be set on the <httpRuntime> tag.
+      <system.Web>
+        <httpRuntime targetFramework="4.5.1" />
```
Comment (thuannguy) [inline Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config]: 4.5.1 or 4.8?

File: Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config [hunk 12]
```diff
@@ -65,11 +67,14 @@
       </md:ContactPerson>
     </ServiceProvider>
     <RequestedAttributes>
-      <att name="dk:gov:saml:attribute:AssuranceLevel" isRequired="true" />
-      <att name="dk:gov:saml:attribute:SpecVer" isRequired="true" />
+      <att name="https://data.gov.dk/model/core/specVersion" isRequired="true" />
+      <att name="https://data.gov.dk/model/core/eid/privilegesIntermediate" isRequired="true" />
+      <att name="https://data.gov.dk/concept/core/nsis/loa" isRequired="true" />
+      <att name="https://data.gov.dk/model/core/eid/email" isRequired="true" />
+      <att name="https://data.gov.dk/model/core/eid/cprNumber" isRequired="true" />
```
Comment (thuannguy) [inline Deployment/Kombit.Samples.CH.WebsiteDemo/Web.config]: CPR is not required

**Review by thuannguy — CHANGES_REQUESTED** [review]

**Review by thuannguy — APPROVED** [review]
